### PR TITLE
Complete repo refactor

### LINKED ISSUE #24: Repo wide refactor
Right now the repo is a giant mess, adding anything is very hard because the code isn't modular, that would've worked if it were a small project, but this one's getting a bit out of hand.

### PR DESCRIPTION
**Pull request by KyberNull**
This pull request introduces a comprehensive configuration system, modularizes dataset and processing code, and improves model flexibility and maintainability. The main changes are the introduction of dedicated configuration files for each pipeline stage, the consolidation and renaming of dataset and transform utilities under a new `processing` package, and the enhancement of the SegFormer model to support gradient checkpointing. Additionally, the training and evaluation scripts are refactored to use the new configuration and processing modules, improving code clarity and reusability.

This PR  Fixes #24.

**Configuration system and code modularization:**

* Added a `config` package with shared and stage-specific configuration files (`shared.py`, `eval.py`, `inference.py`, `phase2.py`, `phase3.py`), and updated all scripts to load their parameters from these modules. [[1]](diffhunk://#diff-e192b8eb2490fd56e24eb7a2e273ac382bb1ddf3dc7cb259f630281e2772f213R1-R25) [[2]](diffhunk://#diff-e86b1b1d70f7b80727751ec5f3b0481f2c497bb0bce2ae4f29bb106e0304d659R1-R8) [[3]](diffhunk://#diff-f6deb312ba0ee31ee3c0fae3a585d396af3530a3e495a44cb8d89969ff1802c1R1-R8) [[4]](diffhunk://#diff-05e94dc8714e7a939f221b4fac18ab997bdf29bc2e1d368389085fbc8aedc204R1-R7) [[5]](diffhunk://#diff-5f35b7a4a648669823b7ac8a50968cbe40d6d6bc97ffc62fe5ad4db48386afc6R1-R11) [[6]](diffhunk://#diff-e714b16a0e92f02d6fbea552cc6f242bf34fc7af96190eddd0602ce50a16a146R1-R11) [[7]](diffhunk://#diff-24edcdcba3747a424b769b374189e8346f6c2f7bdb1e16d2d9ffbe9b926d45a7L3-R11) [[8]](diffhunk://#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R1-R9)
* Introduced a new `processing` package consolidating dataset loading, augmentations, and postprocessing. The old `datasets.py` was renamed to `processing/dataset.py`, and transforms/postprocessing utilities are now imported from `processing`. [[1]](diffhunk://#diff-5bd58c2e615d33fcb47427cb3a9af94f1596d56f00a69eb4b1b75dd5e6f006aeR1-R23) [[2]](diffhunk://#diff-54f00a84123c3596e6ad23969b7c6990b4302b6015c084a0dd0222120b035cd9L1-L33) [[3]](diffhunk://#diff-54f00a84123c3596e6ad23969b7c6990b4302b6015c084a0dd0222120b035cd9L43) [[4]](diffhunk://#diff-24edcdcba3747a424b769b374189e8346f6c2f7bdb1e16d2d9ffbe9b926d45a7L15-R37) [[5]](diffhunk://#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1L17-R26)

**Model improvements:**

* Enhanced the `SegFormer` model in `model.py` to support gradient checkpointing for memory efficiency, with fallback logic if the encoder does not support it. [[1]](diffhunk://#diff-fada037ad086638e65c7ae77e3d223963e9afaa26326aab0ea718f4013176e43L6-R31) [[2]](diffhunk://#diff-fada037ad086638e65c7ae77e3d223963e9afaa26326aab0ea718f4013176e43L35-R48)

**Training and evaluation script refactoring:**

* Refactored `evaluate.py` and `main.py` to use the new configuration and processing modules, removing hardcoded constants and improving maintainability. [[1]](diffhunk://#diff-24edcdcba3747a424b769b374189e8346f6c2f7bdb1e16d2d9ffbe9b926d45a7L3-R11) [[2]](diffhunk://#diff-24edcdcba3747a424b769b374189e8346f6c2f7bdb1e16d2d9ffbe9b926d45a7L15-R37) [[3]](diffhunk://#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R1-R9) [[4]](diffhunk://#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1L17-R26) [[5]](diffhunk://#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1L33-R46)
* Simplified `pretrain.py` to act as a wrapper that delegates to `training.train.main`, removing duplicated training logic and centralizing it in a single location.

These changes collectively make the codebase more modular, configurable, and easier to maintain or extend for future development.

### CODE REVIEW

**Review by O-tech477 — DISMISSED** [review]
LGTM !! Big PR ngl

**Comment by KyberNull** [comment]
Check it by running it.

File: processing/__init__.py
Comment (KyberNull) [inline processing/__init__.py:1]: Why send them up?

File: processing/postprocessing.py [hunk 1]
```diff
@@ -11,7 +11,8 @@
 class PostProcessing:
     def __init__(self, num_classes):
         self.num_classes = num_classes
-        self.processors = {1: PostProcessingWater(), 2: PostProcessingBuildings(), 3: PostProcessingRoads()}
+        # Global class ids: 0 background, 1 roads, 2 buildings, 3 water.
+        self.processors = {1: PostProcessingRoads(), 2: PostProcessingBuildings(), 3: PostProcessingWater()}
```
Comment (KyberNull) [inline processing/postprocessing.py:15]: This will cause roads to be over-ridden by buildings and water, is that intended?

**Review by O-tech477 — DISMISSED** [review]
LGTM! Working

File: processing/postprocessing.py (hunk 1 above)
Comment (O-tech477) [reply processing/postprocessing.py:15]: the roads have the class index of 1, so for roads PostProcessingRoads must be called ?

File: processing/__init__.py
Comment (O-tech477) [reply processing/__init__.py:1]: removes circular dependencies of trasnforms getting called again in __init__.py

**Review by O-tech477 — APPROVED** [review]
LGTM!!
